en/lessons/basics/functions.md

```diff
@@ -1,425 +1,83 @@
-
-```elixir
-iex> handle_result = fn
-...>   {:ok, result} -> IO.puts "Handling result..."
-...>   {:ok, _} -> IO.puts "This would be never run as previous will be matched beforehand."
-...>   {:error} -> IO.puts "An error has occurred!"
-...> end
-
-iex> some_result = 1
-1
-iex> handle_result.({:ok, some_result})
-Handling result...
-:ok
-iex> handle_result.({:error})
-An error has occurred!
+   \x -> x
+    │ │  └─ Body of the lambda
+    │ └─ Separator between head and body of the lambda
+    └─ Declaration of the input bindigs
 ```
 
-## Named Functions
-
-We can define functions with names so we can easily refer to them later.
-Named functions are defined within a module using the `def` keyword .
-We'll learn more about Modules in the next lessons, for now we'll focus on the named functions alone.
+And incidentally, we have also seen one of the most primordial functions : The identity function, or `id`.  
+This function is very simple, as it returns precisely the argument that we pass to it.
 
-Functions defined within a module are available to other modules for use.
-This is a particularly useful building block in Elixir:
-
-```elixir
-defmodule Greeter do
-  def hello(name) do
-    "Hello, " <> name
-  end
-end
-
-iex> Greeter.hello("Sean")
-"Hello, Sean"
-```
+## Named functions
 
-If our function body only spans one line, we can shorten it further with `do:`:
+However, one can only go so far with anonymous functions. This is why we can give names to our functions:
 
-```elixir
-defmodule Greeter do
-  def hello(name), do: "Hello, " <> name
-end
-```
-
-Armed with our knowledge of pattern matching, let's explore recursion using named functions:
-
-```elixir
-defmodule Length do
-  def of([]), do: 0
-  def of([_ | tail]), do: 1 + of(tail)
-end
-
-iex> Length.of []
-0
-iex> Length.of [1, 2, 3]
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> addOne x = x + 1
+ghci> addOne 2
 3
 ```
 
-### Function Naming and Arity
+## Higher-order functions
 
-We mentioned earlier that functions are named by the combination of given name and arity (number of arguments).
-This means you can do things like this:
+We can now see a third way of using functions, which is the composition of both anonymous and named functions:
+Higher-order functions!
 
-```elixir
-defmodule Greeter2 do
-  def hello(), do: "Hello, anonymous person!"   # hello/0
-  def hello(name), do: "Hello, " <> name        # hello/1
-  def hello(name1, name2), do: "Hello, #{name1} and #{name2}"
-                                                # hello/2
-end
+They become particularly useful when you need to control the application of a function over a parameter:
 
-iex> Greeter2.hello()
-"Hello, anonymous person!"
-iex> Greeter2.hello("Fred")
-"Hello, Fred"
-iex> Greeter2.hello("Fred", "Jane")
-"Hello, Fred and Jane"
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> apply2 f x = f (f x)
```

Comment:
Perhaps this function should be called `applyTwice`. I think it would be more consistent with the other function names (like `addOne`) in this document.

Reply:
yep